tests/integration/ha_tests/test_self_healing_2.py

```diff
@@ -0,0 +1,204 @@
+#!/usr/bin/env python3
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+import asyncio
+import logging
+from time import sleep
+
+import pytest
+from pytest_operator.plugin import OpsTest
+from tenacity import Retrying, stop_after_delay, wait_fixed
+
+from ..helpers import (
+    CHARM_BASE,
+    get_password,
+)
+from .conftest import APPLICATION_NAME
+from .helpers import (
+    METADATA,
+    ORIGINAL_RESTART_CONDITION,
+    app_name,
+    are_all_db_processes_down,
+    are_writes_increasing,
+    change_patroni_setting,
+    check_writes,
+    fetch_cluster_members,
+    get_patroni_setting,
+    get_primary,
+    is_cluster_updated,
+    is_postgresql_ready,
+    send_signal_to_process,
+    start_continuous_writes,
+    update_restart_condition,
+)
+
+logger = logging.getLogger(__name__)
+
+APP_NAME = METADATA["name"]
+PATRONI_PROCESS = "/snap/charmed-postgresql/[0-9]*/usr/bin/patroni"
+POSTGRESQL_PROCESS = "postgres"
+DB_PROCESSES = [POSTGRESQL_PROCESS, PATRONI_PROCESS]
+MEDIAN_ELECTION_TIME = 10
+
+
+@pytest.mark.abort_on_fail
+async def test_build_and_deploy(ops_test: OpsTest, charm) -> None:
+    """Build and deploy three unit of PostgreSQL."""
+    wait_for_apps = False
+    # It is possible for users to provide their own cluster for HA testing. Hence, check if there
+    # is a pre-existing cluster.
+    if not await app_name(ops_test):
+        wait_for_apps = True
+        async with ops_test.fast_forward():
+            await ops_test.model.deploy(
+                charm,
+                num_units=3,
+                base=CHARM_BASE,
+                storage={
+                    "archive": {"pool": "lxd-btrfs", "size": 2048},
+                    "data": {"pool": "lxd-btrfs", "size": 2048},
+                    "logs": {"pool": "lxd-btrfs", "size": 2048},
+                    "temp": {"pool": "lxd-btrfs", "size": 2048},
+                },
```

Comment:
We only need the storage for `test_storage_re_use`.

Reply:
That's a good catch. Thanks, Dragomir! I'll remove it in the next PR to unblock the other PRs.